### PR TITLE
WIP: Add GPU e2e_node test for checking kubelet restarts

### PR DESCRIPTION
**Pull request by cmluciano**
GPUs are initialized at startup and should recover when the kubelet
is restarted.

Signed-off-by: Christopher M. Luciano <cmluciano@us.ibm.com>

Related #46087

**Special notes for your reviewer**:
Marked as WIP since I cannot test locally and probably made some mistakes

**Release note**:
None

### CODE REVIEW

File: test/e2e_node/gpus.go [hunk 1]
```diff
@@ -151,6 +151,29 @@ var _ = framework.KubeDescribe("GPU [Serial]", func() {
 
 			By("Checking if the pod outputted Success to its logs")
 			framework.ExpectNoError(f.PodClient().MatchContainerOutput(podFailure.Name, podFailure.Name, "Success"))
+
+			By("checking allocatable gpus")
+			gpusAvailableBefore := getGPUsAvailable(f)
+
+			By("restarting the kubelet")
+			if stdout, err := exec.Command("sudo", "systemctl", "restart", "kubelet").CombinedOutput(); err != nil {
```
Comment (vishh) [inline test/e2e_node/gpus.go]: Why not `sudo pkill kubelet` and that way it will work even without systemd?

File: test/e2e_node/gpus.go [hunk 2]
```diff
@@ -151,6 +151,29 @@ var _ = framework.KubeDescribe("GPU [Serial]", func() {
 
 			By("Checking if the pod outputted Success to its logs")
 			framework.ExpectNoError(f.PodClient().MatchContainerOutput(podFailure.Name, podFailure.Name, "Success"))
+
+			By("checking allocatable gpus")
+			gpusAvailableBefore := getGPUsAvailable(f)
```
Comment (vishh) [inline test/e2e_node/gpus.go]: `Available` here refers to `Capacity/Allocatable`. What you need I assume is the number of GPUs that are free right?

File: test/e2e_node/gpus.go [hunk 3]
```diff
@@ -151,6 +151,29 @@ var _ = framework.KubeDescribe("GPU [Serial]", func() {
...
+			if stdout, err := exec.Command("sudo", "systemctl", "restart", "kubelet").CombinedOutput(); err != nil {
+				framework.Logf("Failed to trigger kubelet restart with systemd/systemctl: %v, stdout: %q", err, string(stdout))
+				if stdout, err = exec.Command("sudo", "service", "kubelet", "restart").CombinedOutput(); err != nil {
+					framework.Failf("Failed to trigger kubelet restart with upstart/service: %v, stdout: %q", err, string(stdout))
+				}
+			}
+			time.Sleep(20 * time.Second)
+
+			By("expecting the same number of allocatable gpus")
+			gpusAvailableAfter := getGPUsAvailable(f)
```
Comment (vishh) [inline test/e2e_node/gpus.go]: I don't think this PR achieves your intent which IIUC is that of testing if kubelet keeps track of allocated GPUs across restarts.
Why not, schedule a pod that consumes all GPUs, then restart kubelet, and then try scheduling another pod that wants a GPU and validate that the new pod stays pending?

File: test/e2e_node/gpus.go (hunk 2 above)
Comment (cmluciano) [inline test/e2e_node/gpus.go]: Yes that would work. What I was hoping to do was check for any number and make sure that number exists after a kubelet restart.

File: test/e2e_node/gpus.go (hunk 1 above)
Comment (cmluciano) [inline test/e2e_node/gpus.go]: That's a good idea. I copied this code from another test that must be assuming something like Ubuntu.

**Comment by cmluciano** [comment]
@k8s-bot test this

**Comment by cmluciano** [comment]
@vishh Can you take another look here?

**Comment by cmluciano** [comment]
I'm not sure if the kops error is expected or not. cc @kubernetes/sig-aws-misc

**Comment by vishh** [comment]
@cmluciano did you manually test this?

@mindprince if you get a chance can you give this PR a shot on COS base image?

**Comment by cmluciano** [comment]
@vishh I have not manually tested it yet. Is there a bit to flip in order to test it with the COS image?

**Comment by mindprince** [comment]
@cmluciano 
To start these tests from my workstation, I was using:
```
make test-e2e-node REMOTE=true FOCUS="GPU" SKIP="" IMAGE_CONFIG_FILE=image-config.yaml TEST_ARGS="--feature-gates=DynamicKubeletConfig=true"
```
Here's the `image-config.yaml` file:
```
images:
  gci:
    image_regex: cos-beta-59-9460-20-0 # docker 1.11.2
    project: cos-cloud
    metadata: "user-data<test/e2e_node/jenkins/gci-init-gpu.yaml,gci-update-strategy=update_disabled"
    resources:
      accelerators:
        - type: nvidia-tesla-k80
          count: 2
```
This will start the tests in your GCP account.


Make sure to pull in https://github.com/kubernetes/kubernetes/pull/46810 and https://github.com/kubernetes/kubernetes/pull/46899 before running this.

**Comment by cmluciano** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

**Comment by mindprince** [comment]
These tests failed when I started them from my workstation.

**Comment by cmluciano** [comment]
@mindprince weird, is the GPU e2e not run in conjunction with this PR?

**Comment by mindprince** [comment]
@cmluciano Yeah, the GPU e2e node tests don't run on PRs because they are marked "[Serial]".

**Comment by vishh** [comment]
@mindprince a TODO post v1.7 would be that of running GPU e2es per PR. Wanna open an issue for that?

**Comment by dchen1107** [comment]
Move this to v1.8 per above comment.

**Comment by cmluciano** [comment]
@k8s-bot keep-open

**Comment by cmluciano** [comment]
@mindprince @vishh Can I get another review on this? Ie2e is passing https://k8s-gubernator.appspot.com/pr/46644

**Comment by vishh** [comment]
/approve 

This LGTM. I'll let @mindprince lgtm this after he reviews. Thanks @cmluciano

**Comment by mindprince** [comment]
This tests update node_e2e gpu tests and not normal e2e gpu tests. We have started running [normal e2e gpu tests on PRs](https://k8s-testgrid.appspot.com/wg-resource-management#gce-gpu) but we are still not running [node_e2e gpu tests on PRs](https://k8s-testgrid.appspot.com/wg-resource-management#kubelet-serial-gce-e2e).

As my [comment above](https://github.com/kubernetes/kubernetes/pull/46644#issuecomment-306885515) points out, this tests fails the node_e2e gpu tests, we can't merge it.

**Comment by cmluciano** [comment]
closing in favor of #52250

**Pull request closed without merging**